Lib/stdlib_deprecations.py

```diff
@@ -0,0 +1,185 @@
+import dataclasses
+import re
+
+
+@dataclasses.dataclass(slots=True, frozen=True)
+class Deprecated:
+    name: str
+    version: tuple[int]
+    remove: tuple[int] | None
+    message: str | None
+
+
+def _parse_version(version_str):
+    version = tuple(int(part) for part in version_str.split('.'))
+    if len(version) < 2:
+        raise ValueError(f"invalid Python version: {version_str!r}")
+    return version
+
+
+_DEPRECATIONS = {}
+_DEPRECATIONS_CAPI = {}
+_REGEX_NAME = r'[A-Za-z_][A-Za-z0-9_]*'
+_REGEX_QUALNAME = fr'^{_REGEX_NAME}(\.{_REGEX_NAME})*$'
+
+
+def _deprecate_api(api_dict, name, version, remove, replace):
+    if not re.match(_REGEX_QUALNAME, name):
+        raise ValueError(f"invalid name: {name!a}")
+    version = _parse_version(version)
+    if remove is not None:
+        remove = _parse_version(remove)
+    if replace is not None:
+        msg = f'use {replace}'
+    else:
+        msg = None
+    deprecated = Deprecated(name, version, remove, msg)
+    api_dict[deprecated.name] = deprecated
+
+
+def _deprecate(name, version, *, remove=None, replace=None):
+    global _DEPRECATIONS
+    _deprecate_api(_DEPRECATIONS, name, version, remove, replace)
+
+
+def _deprecate_capi(name, version, *, remove=None, replace=None):
+    global _DEPRECATIONS_CAPI
+    _deprecate_api(_DEPRECATIONS_CAPI, name, version, remove, replace)
+
+
+# Python 2.6
+_deprecate('gzip.GzipFile.filename', '2.6', remove='3.12',
+    replace='gzip.GzipFile.name'),
+
+# Python 3.6
+_deprecate('asyncore', '3.6', remove='3.12', replace='asyncio'),
+_deprecate('asynchat', '3.6', remove='3.12', replace='asyncio'),
+_deprecate('smtpd', '3.6', remove='3.12', replace='aiosmtp'),
+_deprecate('ssl.RAND_pseudo_bytes', '3.6', remove='3.12',
+    replace='os.urandom()'),
```

Comment:
Does this mean the exact call `os.urandom()` without args, or is it a reference to the function with parentheses added like in C documentation?

Being a Python dev and not a C dev, I take `os.urandom` to mean one thing and `os.urandom()` to mean another thing!

Reply:
os.urandom() has no arguments. Adding parenthesis is a hint to show that it's a function

Reply:
OK but my point is that it’s not a great doc convention for Python.  I think the functions that follow need arguments but have `()` too, that’s misleading.

Reply:
I presume that `ssl.RAND_pseudo_bytes` is the name of a function.  The exact replacement is `os.random`, without `()`.  I think it should be given than way.

Reply:
In practice, I expect that the most usual usage of a function like ssl.RAND_pseudo_bytes is to call it: ``ssl.RAND_pseudo_bytes()``. So I prefer to specify the replacement with parenthesis as well: ``os.urandom()``. In my current implementation, Deprecation.message is an arbitrary string, it's not designed to automate replacement by a linter or a similar tool.